### ISSUE TITLE
Firmware update does not work

### ISSUE THREAD

**Issue by fdanapfel** [rabbitholecomputing/SCSI2SD-util-V6#11]
Hello,

I've just tried using SCSI2SD-util-v6 to update the firmware of a SCSI2SD v6/2020d board using the firmware.V6.2020.dfu file from http://www.codesrc.com/files/scsi2sd-v6/latest/, but when I select the 'Update firmware...' menu item and select the firmware file a dialog box pops up stating 'Wrong filename Firmware does not match device hardware' and nothing more happens after I close the dialog box by clicking OK. I have to close the app, disconnect the SCSI2SD V6/2020d card and then restart the app again and reconnect the cart to be able to configure the card.

The log window shows the following:
```
SCSI2SD Ready, firmware version 6.3.1
Hardware version: V6, 2020c or newer
Serial Number: 366B38693138
SD Capacity (512-byte sectors): 31116288
SD CSD Register: 0000000000000000
SD CID Register: 100E032B5590000000
SCSI2SD Ready, firmware version 6.3.1
Hardware version: V6, 2020c or newer
Serial Number: 366B38693138
SD Capacity (512-byte sectors): 31116288
SD CSD Register: 0000000000000000
SD CID Register: 100E032B5590000000

Load config settings
Reading sector 31116286
Reading sector 31116287
Load Complete
Firmware does not match hardware
```

I'm using the current version of SCSI2SD-util-v6 available from https://github.com/rabbitholecomputing/SCSI2SD-util-V6/releases/tag/1.0.6

**Comment by aperezbios** [comment]
@fdanapfel , thanks for the report, and for providing the output from the log window. Not that it should matter, but is this running on M1 Apple silicon, or is it an Intel-based machine?

We'll try to replicate it here.

**Comment by fdanapfel** [comment]
@aperezbios Thanks for taking a look at this.  I'm trying this on an Intel-based machine (late 2013 iMac 27") running macOS Catalina (10.15.7).

**Comment by fdanapfel** [comment]
I just noticed another issue related to this: the "Upgrade Firmware ..." menu item is also active when no SCSI2SD card is attached. When selecting it without having a card attached it still prompts for the firmware file, the "DFU update" window pops up (but is "backgrounded" immediately) but then nothing else happens afterward, and SCSI2SD-util-v6 has to be restarted in order to be useable again.

It would therefore be better if the "Upgrade Firmware ..." menu item  was only selectable when a card has actually been detected by the App.

**Comment by aperezbios** [comment]
Unfortunately it's an entirely valid thing to do, because if there's no firmware on the board, you still need to be able to re-flash the firmware, and the utility can't detect the board if there's no functional firmware on it. Going to initiate a firmware "update" on a board you're re-flashing is something people occasionally need to do, as a form of rescue.

**Comment by fdanapfel** [comment]
Maybe you could then at least add a dialog box that says something like "No SCSI2SD board detected. Do you really want to Update/install firmware? Yes/No".

But this is low priority, I still would like to get the original issue of not being able to actually update the firmware resolved first.

**Comment by aperezbios** [comment]
@fdanapfel can you attempt a firmware upgrade with this version of the utility? http://www.codesrc.com/files/scsi2sd-v6/v6.4.12/mac/scsi2sd-util6.dmg

It isn't signed/notarized, so you would have to override the default security settings (in the preferences pane) to be able to execute it. It doesn't work properly with dark mode, because it's based on an old version of wxWindows. It would be helpful to know if this allows you to upgrade successfully.

**Comment by fdanapfel** [comment]
@aperezbios Updating the firmware using  http://www.codesrc.com/files/scsi2sd-v6/v6.4.12/mac/scsi2sd-util6.dmg is working, using exactly the same firmware file:

20:42:07: SCSI2SD Ready, firmware version 6.3.1
20:42:07: Hardware version: V6, 2020c or newer
Serial Number: 366B38693138
20:42:07: SD Capacity (512-byte sectors): 31116288
SD CSD Register: 0100000e0032b5050900000000000000
SD CID Register: 03004453363143534780000077b8b6e2
20:42:33: Attempting firmware update from file /Users/frank/Downloads/SCSI2SD/firmware/firmware.V6.2020.dfu
20:42:33: Searching for bootloader
20:42:33: Resetting SCSI2SD into bootloader
20:42:34: USB HID read failure: 0x0
20:42:34: STM DFU Bootloader found
20:42:34: Running: dfu-util —download “/Users/frank/Downloads/SCSI2SD/firmware/firmware.V6.2020.dfu” —alt 0 —reset
20:42:34: Debug: wxMacLaunch Bad bundle: dfu-util
20:42:59: Firmware update succeeded
20:44:52: SCSI2SD Ready, firmware version 6.4.12
20:44:52: Hardware version: V6, 2020c or newer
Serial Number: 366B38693138
20:44:52: SD Capacity (512-byte sectors): 31116288
SD CSD Register: 32000e400000595b807fb2761240400a
SD CID Register: 5344530347363143b8b6e280c6460177

**Comment by aperezbios** [comment]
Thanks for the update here. That will help narrow things down nicely.

**Comment by zezba9000** [comment]
I have the same Problem on Windows 11.
SCSI2SD v6 2021b

And its not just firmware issues. 50% of the time the app hangs. Had to plug in device quick & save settings before the app hung. Or I wouldn't even be able to use the device. v5 was much more stable.

**Comment by aperezbios** [comment]
@zezba9000 on Windows, you must have the DFU-util driver installed. Have you installed it? This bug is for the _native mac version_ of SCSI2SD-util-v6.  It sounds like you may have multiple instances of scsi2sd-util6 open. Please e-mail Michael McMaster about this issue, he needs to know. He wrote the utility.

**Comment by zezba9000** [comment]
Yes drivers are installed "UsbDk_1.0.21_x64.msi" & using app version "scsi2sd-util6-x64-6.4.11"
It can communicate with device but only for a little bit.
The error on macOS is the same as Windows 11 "Firmware does not match device hardware".

**I just emailed him to let him know about it**. At least I got the device working after many tries but I'd like to update firmware.

**Comment by fdanapfel** [comment]
@aperezbios FYI. I just tried using version 1.0.8 of SCSI2SD-util-v6 and I'm still getting the same issue when trying to update the firmware of a SCSI2SD v6 2020d card using the official firmware file from http://www.codesrc.com/files/scsi2sd-v6/latest/

Are there any plans to still try to get this fixed in your version of the SCSI2SD utility for v6 cards, or is using the official tool  from codescr the only option to update the firmware on the cards on macOS 10.15.7 on x86_64?

**Comment by zezba9000** [comment]
v5 of SCSI2SD doesn't have this issue.
For such an expensive board I don't know why this isn't documented.

The response I got back was
```
Try using https://zadig.akeo.ie/ to install the “WinUSB” driver for the STM32 BOOTLOADER device.
… dfu-util is using a different driver (usbdk), but it can have trouble if there’s no loaded driver available to redirect the device from.
```

I didn't try & just gave up as it was talking to long and I needed to just finish putting back together the computer.
Maybe that will work for you @fdanapfel ?

I'm also finding BlueSCSI to just be so much easier to use as it actually supports disk images in Fat32 or FatEx file systems.

**Comment by nochristrequired** [comment]
@zezba9000 you made a comment about using Zadig to install the WinUSB driver. This is what I had to do on Windows 10. My device actually bricked itself with the correct firmware file when I tried to update prior to doing this. There's no mention on the wiki on how to recover the device or anything. Very frustrating but I did manage to recover it by shorting the BOOTLDR and using your advice to reflash it.

**Comment by kskeldon** [comment]
Hi there, I just bought a SCSD2SD V6 / 2021 card to use with my kurzweil sampler. Alas I seem to have made the firmware bad by trying to flash settings to it. No matter what firmware upgrade I try, it gives me a "firmware does not match device" error. Or, of I try to upgrade it using the bootloader method (selecting file before connecting usb) it just gives an "update failed" message. Can anyone help? I am not sure even which dfu file to use, so have been trying all of them! I'm using Windows10. Thanks.

**Comment by zezba9000** [comment]
@kskeldon

This is the response I got from developer. I gave up and just used firmware that came on device. (got lucky and settings stuck though).
```
Try using https://zadig.akeo.ie/ to install the “WinUSB” driver for the STM32 BOOTLOADER device.
… dfu-util is using a different driver (usbdk), but it can have trouble if there’s no loaded driver available to redirect the device from.
```

**Comment by kskeldon** [comment]
Many thanks for this. I was about to try zadig,  however, although I don't quite know how, I got it working. I then had issues getting windows to see the 4 drives configured in the SCSI2SD firmware and subsequently formatted in the sampler, even though they were FAT. But doing another simple FAT format on the PC seemed to allow both kurzweil and PC to access the 4 x drives.
Generally I'm finding I have to exercise any load/write command in the V6 utility app a few times to be sure it works (even if it seems from the log it has). And also have the SD card fitted, even though it shouldn't be needed for config writes to the firmware. But when actually configured and all is done,  it seems pretty stable (so far).

**Comment by zezba9000** [comment]
TBH for how expensive this device is and the lack of docs for v6. I'd just stick with SCSI2SD v5 or use BlueSCSI.
I've had zero issues with those.

**Comment by aperezbios** [comment]
There are many thousands of SCSI2SD V6's in use with zero problems. SCSI2SD V6 and BlueSCSI aren't anywhere even close to comparable, performance-wise. Furthermore, this is an issue ticket for SCSI2SD V6, and not anything else. Please stay on-topic. 

@zezba9000 What documentation are you in need of, specifically?

**Comment by aperezbios** [comment]
> Generally I'm finding I have to exercise any load/write command in the V6 utility app a few times to be sure it works (even if it seems from the log it has). And also have the SD card fitted, even though it shouldn't be needed for config writes to the firmware. But when actually configured and all is done, it seems pretty stable (so far).

With V6, the SD card actually MUST be inserted for config saves and reads to be successful, because the configuration metadata lives in the last two sectors/blocks of the SD card itself, not internally within the flash. This is a change of behavior which was introduced when V6 was originally developed and released, which was in May of 2016, nearly six years ago. 

If you're having trouble reliably saving and reading configurations to your V6, try with a different, known-good USB cable. Possibly a shorter one. There are a lot of extremely low quality USB-A to microUSB cables out there, and iin our experience, the microUSB connector itself, on the cable, is often actually the problem.

**Comment by zezba9000** [comment]
> @zezba9000 What documentation are you in need of, specifically?

The required driver docs for Windows don't exist outside me mailing the developer of the device that I could find & they don't exist in this GitHubs readme or in the firmwares zip readme.

> If you're having trouble reliably saving and reading configurations to your V6, try with a different, known-good USB cable.

This is not the issue with v6 & I have very good USB cables I use for Android development. The drivers for Windows instructions are documented incorrectly or only work for old Windows computers (which is why some of us are here). They are completely unstable with Win10 / Win11 at least.

If these 'zadig' driver requirements are documented somewhere outside this GitHub issue I hope they are migrated and included in the firmware readme file that many of us are likely looking at.
Maybe both drivers can be shipped in zip firmware file if there isn't a license issue. At min readme should link to them.

**Comment by aperezbios** [comment]
They absolutely do work in most cases, just not yours. 

Since you mention Android application development, this is a possible, even likely, cause of the issues you've encountered, as it's extremely likely that you have/had another, pre-existing DFU driver installed. This is a known corner-case, which the Zadig utility can be used to resolve. Just because something doesn't work in one specific instance (yours, in this case) it's not logical to conclude that it doesn't work in any others. While I can appreciate that it's human nature to assume this must be the case (it not working for you, therefore it must not work for anyone), this is a logical fallacy. We'd be happy to help you work through your specific issues and circumstances. Have you contacted the manufacturer/reseller of your device?

**Comment by aperezbios** [comment]
> This is not the issue with v6 & I have very good USB cables I use for Android development. The drivers for Windows instructions are documented incorrectly or only work for old Windows computers (which is why some of us are here). They are completely unstable with Win10 / Win11 at least.

The Device Firmware Update drivers for SCSI2SD V6 are known to work just fine with 64-bit versions of Windows 7, 10, and 11, as well as under macOS. If you are experiencing a circumstance where that's not the case, we encourage you to provide specific steps to replicate it, if possible. We recognize that this is not always easy or straightforward, but being able to reproduce a bug is necessary, if it's to ever be fixed. Without knowing what, exactly, you did or did not do, prior to a failed attempt at upgrading the firmware, it's not possible to provide effective support, when it comes to issues such as this.

**Comment by zezba9000** [comment]
> it's not possible to provide effective support,

If the issue is known & so is the solution then as I've stated multiple times I'd just update the readme / docs in a "known issues" section or something. We're clearly both head-strong but I was only trying to help others based on my experience.

> They absolutely do work in most cases, just not yours....
> ...Just because something doesn't work in one specific instance (yours, in this case) it's not logical to conclude that it doesn't work in any others.

I'm clearly not the only one having this issue & thus I wasn't making a logical fallacy by assuming anything outside reason. I'm not trying to start a war here. I'm just sharing my experience that regardless of its rarity may be of value to others in the same boat.

I'm glad their is a solution to this but again I'd urge it be added to "known issues" in the readme. This is a great example of the confusion it missing can cause. And I do mean the best.

**Comment by akedani** [comment]
Hi,

I have managed to brick my SCSI2SD trying to update the firmware. Is there a method to bring the board back to life?

Gr,
Åke Danielson

**Comment by aperezbios** [comment]
@akedani you need to specify which hardware version of SCSI2SD you're using before we can offer any meaningful assistance.

**Comment by akedani** [comment]
Sorry,

I have a SCSI2SD V6 connected to an Akai S6000. Windows 11 user.

Gr,
Åke

**Comment by aperezbios** [comment]
Åke, 

This is an entirely different problem than the one that was reported here, so this isn't really a suitable place to discuss it, but you're going to need to use dfu-util to re-flash the board. The process for doing that is documented here http://www.codesrc.com/files/scsi2sd-v6/latest/SCSI2SD_QuickStartGuide.pdf

Please reach out to the e-mail address on your SCSI2SD card if you need further guidance.

**Comment by akedani** [comment]
Thanks a million. I'll try i tomorrow!!!

Åke

**Comment by EditLater** [comment]
Hello can you or someone please Help! I am trying to set up A SCSI2SD Zulu V6 rev 2022d card with my Akai s3000xl. I am using the scsi2sd-util6 I have installed the UsbDk driver and the dfu-util. I have tried all of the v6 firmware updates from codesrc.com and all I get is "Searching for bootloader". My USB cable is good. I've not been connecting the Usb cable until it starts searching, as per instructions . I've tried with the sampler on and off, But no Joy. If anybody has any insights I would much appreciate.

**Comment by fdanapfel** [comment]
Looks like no more development is happening on this tool, therefore I'm closing this issue.
